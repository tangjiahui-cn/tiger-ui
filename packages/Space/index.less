@import "../_style/theme";

.space {
  display: inline-flex;
  align-items: center;

  &-wrap {
    flex-wrap: wrap;
  }

  &-block {
    display: flex;
  }

  &-vertical {
    flex-direction: column;
  }
}
