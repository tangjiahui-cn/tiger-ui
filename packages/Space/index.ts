/**
 * Space
 *
 * @author tangjiahui
 * @date 2024/4/23
 */
import Space from './space';

export default Space;
export type * from './space';
