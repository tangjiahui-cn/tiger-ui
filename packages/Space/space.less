@import '../_theme/variable.less';

@prefix: prefix('space');

@{prefix} {
  display: inline-flex;

  &-wrap {
    flex-wrap: wrap;
  }

  &-vertical {
    flex-direction: column;
  }

  &-horizontal {
    align-items: center;
  }

  &-block {
    display: block;
  }
}
