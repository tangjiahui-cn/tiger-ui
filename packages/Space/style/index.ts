import usePrefix from '../../_utils/hooks/usePrefix';
import useEmotion from '../../_utils/hooks/useEmotion';
import brotherSelector from '../../_utils/style/brotherSelector';
import type { SpaceDirection } from '..';

type StyleObject = {
  [k: string]: any | StyleObject;
};

export function useStyle(componentName: string): {
  space: () => string;
  wrap: () => string;
  block: () => string;
  direction: (direction: SpaceDirection) => string;
} {
  const prefix = usePrefix(componentName);
  const { css } = useEmotion({ key: prefix });

  const wrap = () => `${prefix}-wrap`;
  const block = () => `${prefix}-block`;

  function genDirectionStyle() {
    const getClassName = (direction: SpaceDirection) => `${prefix}-${direction}`;
    const directionStyle = {
      [getClassName('vertical')]: {
        flexDirection: 'column',
      },
      [getClassName('horizontal')]: {
        alignItems: 'center',
      },
    };
    return {
      direction: getClassName,
      directionStyle,
    };
  }

  const { direction, directionStyle } = genDirectionStyle();

  const space = () => {
    return css({
      display: 'inline-flex',
      ...brotherSelector(directionStyle),
      ...brotherSelector({
        [wrap()]: {
          flexWrap: 'wrap',
        },
      }),
      ...brotherSelector({
        [block()]: {
          display: 'flex',
        },
      }),
    });
  };

  return {
    space,
    wrap,
    block,
    direction,
  };
}
