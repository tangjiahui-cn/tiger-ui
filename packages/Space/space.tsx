/**
 * Space
 *
 * @author tangjiahui
 * @date 2023/5/1
 */

import React, { DOMAttributes, ForwardedRef, RefAttributes, useMemo } from 'react';
import classNames from 'classnames';
import { usePrefix } from '@/ConfigProvider/ConfigProvider';
import './space.less';

export type SpaceDirection = 'vertical' | 'horizontal';
export interface BaseSpaceProps {
  /** gap */
  size?: number;
  /** display direction */
  direction?: SpaceDirection;
  /** block status */
  block?: boolean;
  /** auto wrap */
  wrap?: boolean;
  /** space item className */
  itemClassName?: string;
  /** space item style */
  itemStyle?: React.CSSProperties;
  /** space style */
  style?: React.CSSProperties;
  /** space children */
  children?: React.ReactNode[] | React.ReactNode;
  /** space className */
  className?: string;
}

export type SpaceProps = BaseSpaceProps &
  DOMAttributes<HTMLDivElement> &
  RefAttributes<HTMLDivElement>;

export type SpaceFC = React.ForwardRefExoticComponent<SpaceProps>;

function isFragment(el: React.ReactNode): boolean {
  try {
    return (el as React.ReactElement)?.type?.toString?.() === React.Fragment.toString();
  } catch (e) {
    return false;
  }
}

const Space: SpaceFC = React.forwardRef(function (
  props: SpaceProps & DOMAttributes<HTMLDivElement>,
  ref: ForwardedRef<HTMLDivElement>,
) {
  const {
    size = 8,
    direction = 'horizontal',
    block,
    wrap,
    itemClassName,
    itemStyle,
    children,
    style,
    className,
    ...rest
  } = props;

  const prefix = usePrefix('space');

  const innerChildren = useMemo(() => {
    let list: React.ReactNode[] = Array.isArray(children) ? children : children ? [children] : [];

    list = list.reduce((result: React.ReactNode[], x: React.ReactNode) => {
      if (isFragment(x)) {
        result.push(...(x as any)?.props?.children);
      } else if (Array.isArray(x)) {
        result.push(...x);
      } else if (![undefined, null].includes(x as any)) {
        result.push(x);
      }
      return result;
    }, []);
    return list;
  }, [children]);

  const classes = classNames(
    prefix,
    wrap && `${prefix}-wrap`,
    block && `${prefix}-block`,
    `${prefix}-${direction}`,
    className,
  );

  return (
    <div {...rest} ref={ref} className={classes} style={{ gap: size, ...(style || {}) }}>
      {innerChildren?.map((x: React.ReactNode, index) => {
        const key = (x as React.ReactElement)?.key || index;
        return (
          <div key={key} style={itemStyle} className={itemClassName}>
            {x}
          </div>
        );
      })}
    </div>
  );
});

export default Space;
