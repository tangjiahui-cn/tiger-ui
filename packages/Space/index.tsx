import React, { PointerEventHandler, useMemo } from 'react';
import classNames from 'classnames';
import { useStyle } from './style';

export type SpaceDirection = 'vertical' | 'horizontal';

export interface SpaceProps {
  /**
   * 间距大小
   * @description 相邻元素间距像素
   * @default 8
   */
  size?: number;
  /**
   * 排列方向
   * @description 元素排列的方向
   * @default horizontal
   */
  direction?: SpaceDirection;
  /**
   * 是否块元素
   * @description 宽度是否100%
   * @default false
   */
  block?: boolean;
  /**
   * 自动换行
   * @description 是否自动换行（仅在 horizontal 时生效）
   * @default false
   */
  wrap?: boolean;
  /**
   * 容器样式
   * @description 容器内联样式
   */
  style?: React.CSSProperties;
  /**
   * 内部元素
   * @description 包裹元素
   */
  children?: React.ReactNode;
  /**
   * @description 鼠标按下回调
   */
  onPointerDown?: PointerEventHandler<any> | undefined;
}

export default function Space(props: SpaceProps) {
  const style = useStyle('space');

  const classes = classNames(
    style.space(),
    props?.wrap && style.wrap(),
    props?.block && style.block(),
    style.direction(props?.direction || 'horizontal'),
  );

  return (
    <div
      className={classes}
      style={{ gap: props?.size, ...(props?.style || {}) }}
      onPointerDown={props?.onPointerDown}
    >
      {props?.children}
    </div>
  );
}

Space.defaultProps = {
  size: 8,
  wrap: false,
  direction: 'horizontal',
};
