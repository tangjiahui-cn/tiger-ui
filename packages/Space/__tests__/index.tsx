import {
  cssSelectorTest,
  mountTest,
  nullAndUndefinedChildrenTest,
  reactFragmentTest,
  renderTest,
} from '../../../tests';
import Space from '../index';

describe('space', () => {
  renderTest(Space);
  mountTest(Space);
  mountTest(() => (
    <Space>
      1<div>2</div>3
    </Space>
  ));
  mountTest(() => (
    <Space block>
      1<div>2</div>3
    </Space>
  ));
  mountTest(() => (
    <Space direction={'vertical'}>
      1<div>2</div>3
    </Space>
  ));
  mountTest(() => (
    <Space wrap>
      1<div>2</div>3
    </Space>
  ));
  mountTest(() => (
    <Space size={16}>
      1<div>2</div>3
    </Space>
  ));
  mountTest(() => (
    <Space block wrap size={16} direction={'horizontal'}>
      1<div>2</div>3
    </Space>
  ));

  cssSelectorTest(['.space'], Space);
  cssSelectorTest(['.space-wrap'], () => <Space wrap />);
  cssSelectorTest(['.space-vertical'], () => <Space direction={'vertical'} />);
  cssSelectorTest(['.space-block'], () => <Space block />);
  cssSelectorTest(['[style*="gap"]'], () => <Space size={16} />);

  nullAndUndefinedChildrenTest(Space);
  reactFragmentTest(Space);
});
