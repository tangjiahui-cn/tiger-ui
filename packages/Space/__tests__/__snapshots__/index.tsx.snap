// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`space Null and Undefined Children test should render null or undefined in children without errors 1`] = `
<div
  class="space"
  style="gap: 8px;"
/>
`;

exports[`space react fragment test should render React.Fragment without error. 1`] = `
<div
  class="space"
  style="gap: 8px;"
>
  text
</div>
`;

exports[`space render test render correctly 1`] = `
<DocumentFragment>
  <div
    class="space"
    style="gap: 8px;"
  />
</DocumentFragment>
`;
